### ISSUE TITLE
Need some help in understanding the code of TransformerEncoderDecoderModel

### ISSUE THREAD

**Issue by WangHexie** [google-research/pegasus#76]
In the function [ \_\_call\_\_](https://github.com/google-research/pegasus/blob/master/pegasus/models/transformer.py#L76) of the TransformerEncoderDecoderModel.
The following two questions bothers me a lot. 
1. The input is feed into the encoder and saved as `self._context` but it seems not used afterwards.
2. The embedding of target is feed into decoder , then use softmax_cross_entropy to compare the output with the target itself. In my understanding, to minimize to loss, decoder need to ouput the embedding of original target token to get maximum value as  k‖Embedding‖_2. I can't understand the meaning of this.

The [link](https://github.com/google-research/pegasus/blob/master/pegasus/models/transformer.py#L76) to the code.

**Comment by JingqingZ** [comment]
> The input is feed into the encoder and saved as self._context but it seems not used afterwards.

Yes, `self._context` is not actually used but `context` is needed for following computation.

> The embedding of target is feed into decoder , then use softmax_cross_entropy to compare the output with the target itself. In my understanding, to minimize to loss, decoder need to ouput the embedding of original target token to get maximum value as k‖Embedding‖_2. I can't understand the meaning of this.

Sorry, not sure if I understand this question. The decoder outputs predictions (i.e. [logits](https://github.com/google-research/pegasus/blob/939830367bcf411193d2b5eca2f2f90f3f9260ca/pegasus/models/transformer.py#L106)) and compares with the targets to get loss.

**Comment by WangHexie** [comment]
@JingqingZ Thanks for your reply. I missed context input of transformer_block.
And still a few more questions.
Is this the loss of GSG ? If not, then where can I find it? If it is, then the whole sentence(the prediction of  target)  is  generated in one go not like summarization generation since there's no recurrent structure in the __call__ ?

**Comment by JingqingZ** [comment]
The GSG is a sentence selection strategy but not a loss function. In training, the sequence by decoder is generated at once but with upper triangle mask so that the decoder can only cast attention on previous tokens. In testing/inference stage, the sequence is generated by token by token in a loop (see function [predict](https://github.com/google-research/pegasus/blob/939830367bcf411193d2b5eca2f2f90f3f9260ca/pegasus/models/transformer.py#L115)).

**Comment by WangHexie** [comment]
Thanks a lot for your explanation!!! It saved me quite a few time to understand the code, though I still need some time to fully understand it. Nice code and great job!
